File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function (root, subRoot) {
    if (!root) return false;

    return isSame(root, subRoot) || isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
};

var isSame = function (node1, node2) {
    if (!node1 && !node2) return true;
    if (!node1 || !node2) return false;

    return node1.val == node2.val && isSame(node1.left, node2.left) && isSame(node1.right, node2.right)
}

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def hashTree(node):
            if not node: return hash(None)

            leftHash = hashTree(node.left)
            rightHash = hashTree(node.right)

            node.hashVal = hash((node.val, leftHash, rightHash))

            return node.hashVal        

        targetHash = hashTree(subRoot)

        def search(node):
            if not node: return False
            if node.hashVal == targetHash: return True
            
            return search(node.left) or search(node.right)
        
        hashTree(root)
        return search(root)